srcpkgs/zam-plugins/template

```diff
@@ -0,0 +1,103 @@
+# Template file for 'zam-plugins'
+pkgname=zam-plugins
+version=3.14
+revision=1
+build_style=gnu-makefile
+make_use_env=yes
+make_build_args="HAVE_ZITA_CONVOLVER=true"
+hostmakedepends="pkgconf git gendesk"
+makedepends="libX11-devel libglvnd-devel liblo-devel jack-devel ladspa-sdk
+ libsamplerate-devel zita-convolver-devel"
+# Use system zita-convolver instead of the vendored one
+short_desc="LADSPA/LV2/VST/JACK audio plugins for high-quality processing"
+maintainer="Artur Sinila <freesoftware@logarithmus.dev>"
+license="GPL-2.0-or-later"
+homepage="http://zamaudio.com/"
+changelog="https://github.com/zamaudio/zam-plugins/blob/${version}/changelog"
+
+do_fetch() {
+	git clone --depth 1 --branch ${version} --single-branch \
+	--recurse-submodules --shallow-submodules https://github.com/zamaudio/zam-plugins \
+	zam-plugins-${version}
```

Comment:
It should be possible to download tarballs for all submodules or ask them to publish useful tarballs. Cloning a branch name is susceptible to force pushes and you should use commit hashes instead.

If there are too many submodules, reconsider if this is really packageable.

Reply:
> It should be possible to download tarballs for all submodules or ask them to publish useful tarballs. Cloning a branch name is susceptible to force pushes and you should use commit hashes instead.

There isn't any difference between shallow-cloning a tag vs. downloading it in a tarball.
Specially for this case I've just conducted a little experiment to proof that force pushing doesn't alter a commit hash to which tag points to: https://github.com/Logarithmus/test-tag-force-push

> If there are too many submodules, reconsider if this is really packageable.

I don't really see why having submodules is an issue for being packageable. And please don't insist it's a distro policy to have no submodules. There are too many hidden "distro policies", don't you think? Why those policies are nowhere documented? Please, put them into `CONTRIBUTING.md` to avoid further misunderstandings.

AFAIK this package has a single submodule, which is not a problem even if your points were true.

Reply:
> There isn't any difference between shallow-cloning a tag vs. downloading it in a tarball.

Yes there is, we have checksums for the tarball.

Package policy is documented in the Manual; many other packages use submodules by downloading each tarball and moving to the specific place.

> And AFAIK GitHub doesn't allow you to download a tarball for an arbitrary commit, only for a tag.

Did you try? We use this functionality in multiple places.

Anyhow, the issue with being "packageable" is that if writing a reasonable template is too much trouble, then it isn't packageable. It isn't that it has submodules or anything else.

Reply:
> Yes there is, we have checksums for the tarball.

Please, read my message again. You were afraid of force pushes suddenly changing the tag contents. My little experiment proved that tag remains untouched.
**UPD** Now I've came up with this case: author deletes tag completely and then creates it again, but points it to another commit. So maybe you are right that it's better to have a hash in `template` because the package can silently break.

> Package policy is documented in the Manual; many other packages use submodules by downloading each tarball and moving to the specific place.

I don't see using `git` directly instead of tarballs being prohibited in `Manual.md`. Moreover, it documents that `do_fetch` is used to fetch sources when `distfiles` is not set. If you so strongly insist on using tarballs only, then what's the purpose of `do_fetch` then? Let's remove it altogether from `xbps-src` 

> Did you try? We use this functionality in multiple places.

Hmm... OK, looks like `https://github.com/<name>/<repo>/archive/<hash>.tar.gz` works.

> Anyhow, the issue with being "packageable" is that if writing a reasonable template is too much trouble, then it isn't packageable. It isn't that it has submodules or anything else.

There isn't any trouble in writing a template for packages with many submodules. You can just use `do_fetch` and let `git` do it's job, instead of manually copy-pasting commit hashes for each submodule every time the package updates.

Reply:
Yes, I meant retagging as the main issue.

I guess someone has to document these policies, then. They do exist.

`do_fetch` can exist, but it's very much undesired. That a workaround exists for specific templates doesn't mean it should be used all the time.

And we prefer to use submodules explicitly rather than telling git to do things.

Reply:
`distfiles=` are cached by xbps-src and there is no such mechanism for git/do_fetch.

Reply:
@Duncaen thank you for another argument for tarballs